### DISCUSSION TITLE
Fetching data from API endpoint from within the middleware.

### DISCUSSION THREAD

**Discussion by krzysztofRawski** [vercel/next.js#45865] [Help] [answered]
Hi, I'm trying to implement the permanent redirects managed from CMS. I wanted to use a custom endpoint to fetch, cache, and refresh the list of redirects, and then fetch the cached list of redirects in middleware. Simple, and effective. The problem is that fetching data from the custom API endpoint from within the middleware works only in dev. After building the app (yarn build -> yarn start) the fetch throws the error: connect ECONNREFUSED ::1:3000. Does anyone know what is the reason for that and how to combat that issue?

**Comment by icyJoseph** [comment, marked as answer]
Could you share some code? Just to avoid making assumptions.

**Reply by krzysztofRawskiBabylon** [reply]
It can be as simple as that:
`middleware.ts`
```
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export async function middleware(req: NextRequest) {
    const origin = req.nextUrl.origin

    try{
        const data = await fetch(`${origin}/api/example`)
        console.log('Example API output:', await data.json())
    } catch(err: unknown) {
        console.log('Middleware error:', err)
    }

    return NextResponse.next()
}

export const config = {
    matcher: [
        '/((?!api|_next/static|_next/image|favicon.ico|next-assets).*)',
    ],
}
```
and
`pages/api/example.ts`
```
import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(
    req: NextApiRequest,
    res: NextApiResponse
) {
    res.json({message: 'It is Alive'});
}
```

**Reply by icyJoseph** [reply]
I see a server that calls itself, that's usually bound to be brittle, and it's not necessarily how they'd want you to use this API.

I'd check the request header called host instead, but then you need to provide protocol yourself as well, which I guess in dev mode will be http, and in production https, but you gotta put it together yourself.

Some might say that, rather you should be able to abstract away the API route, into reusable functions and make that call directly from the middleware.  

I understand you want to cache the API route response somehow, and try to cut down in transaction time, but in this case the API serving the redirect lists should be one caching that, now necessarily Next.js, it is a bit annoying, I know. 

That is, your CMS should have an endpoint with proper caching that gives the redirect list, to either middleware or API routes.

**Reply by krzysztofRawskiBabylon** [reply]
Well ... middleware should be as fast as it is possible. Sending a request to CMS on every single Next.js request is suboptimal, to say the least (apart from time, any network error leaves you without redirects). 

Moreover having a huge list of redirects, getting them as an array, and filtering them for every request is not cool as well. Using my own API would allow me to optimize CMS requests (you do it once after starting the server, and then you use the webhook to update the list when something changes), it would also allow me to prepare data and provide the middleware with the data structure that would not require loops. 

I can understand that server calling itself is not good, but the lack of any other option to share state between middleware and the rest of the app reduces middleware usage to some more or less trivial use cases (not mentioning other limitations of Edge runtime).

**Reply by krzysztofRawskiBabylon** [reply]
> I'd check the request header called host instead, but then you need to provide protocol yourself as well, which I guess in dev mode will be http, and in production https, but you gotta put it together yourself.

BTW what is the difference in how I build the URL?

**Reply by icyJoseph** [reply]
Aha, I see now, I think:

> Assuming you are not hosting on Vercel, or any of the services that deploy Next.js as a serverless app. In serverless environments API routes are destroyed when not needed.

1. On server start get the data from the CMS, transform it into something that has information for redirecting.
2. When the CMS has an update, it calls an API route, which updates the data on point 1
3. Middleware also gets access to this data by opening an HTTP request, which returns back to the server

It is still not optimal either, but I guess it'll get the job done. 

Did using the hostname work?

**Reply by krzysztofRawskiBabylon** [reply]
> It is still not optimal either, but I guess it'll get the job done.

If there is a better way to solve the problem, I'm open to any suggestions :)

And to answer your question using a hostname did not help.

**Reply by icyJoseph** [reply]
> Ok, to sum up. Calling internal API from Edge runtime middleware will only work on Vercel or any other hosting that deploys middleware as a serverless function

It should work, regardless of the deploy technique. It's just an HTTP transaction, between two servers. 

The problem is getting the correct URL of the API route, it has to be the absolute path. I was hoping that you could get the header HOST. Like in Vercel you could try with VERCEL_URL, and that'll get you want you want.

Like if your user makes a request to `example.com`, the host header should contain that, and hopefully, `example.com/api/redirect` is where your API route lives, so the middleware can make the request to that endpoint, plus the HTTP protocol.

> there is no way to share the state between middleware and the rest of the app

To share JavaScript context between middleware and the rest of the app, is not the way this was meant to be used. It's like, even if you define an object as `globalThis.foo`, it won't be visible to both of them, not even in dev mode I think.

**Reply by krzysztofRawskiBabylon** [reply]
I can hardcode the absolute path and this changes nothing. In dev, it works, and after building the app it throws the error.

**Reply by icyJoseph** [reply]
> If there is a better way to solve the problem, I'm open to any suggestions :)

It is just strange that the host header is not working.

```http
GET http://localhost:3000/api/hello
Host: www.example.com
```

If I send this to my API route, the middleware will intercept it and print `www.example.com`:

```js
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname, href } = req.nextUrl;
  const host = req.headers.get("host");
  console.log(host); // www.example.com

  return NextResponse.next();
}
```

I think you got the right idea, the only issue is figuring out the absolute path to the endpoint.

**Reply by icyJoseph** [reply]
Ah wait a minute! You have IPv6 shape, `::1:3000`, that might be the problem!

I am unfortunately out of time for today, but making a fetch request from middleware to API route should work.

I tested it just now, dev and production, and it worked.

**Reply by krzysztofRawskiBabylon** [reply]
I have disabled the ::1 localhost in my hosts on mac and it works. Do you know why this is a problem on production but not on dev?

**Reply by icyJoseph** [reply]
Absolutely no idea 😅

**Reply by krzysztofRawskiBabylon** [reply]
Is it possible that Next.js production server (yarn start) is IPv4 only?

**Reply by icyJoseph** [reply]
```console
npx next start --help

      Description
        Starts the application in production mode.
        The application should be compiled with `next build` first.

      Usage
        $ next start <dir> -p <port>

      <dir> represents the directory of the Next.js application.
      If no directory is provided, the current directory will be used.

      Options
        --port, -p      A port number on which to start the application
        --hostname, -H  Hostname on which to start the application (default: 0.0.0.0)
        --keepAliveTimeout  Max milliseconds to wait before closing inactive connections
        --help, -h      Displays this message
```

Most likely it is just taking the default, `0.0.0.0`.

**Reply by krzysztofRawskiBabylon** [reply]
Anyway, thanks man. I still don't understand the source of the problem in 100%, but at least I know how to get around it locally and that on production the problem shouldn't occur.

**Reply by icyJoseph** [reply]
Could this be it? https://github.com/node-fetch/node-fetch/issues/1624?

**Reply by krzysztofRawskiBabylon** [reply]
Similar problem, but in my case, I don't use node-fetch inside the middleware I use web standard fetch (I don't know what is being used by edge runtime under the hood). Generally using IP instead of localhost running the server solves the problem.
